base/pointer.jl

```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+function unsafe_store!(p::Ptr, x, i::Integer, order::Symbol)
+    unsafe_store!(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), x, order)
+end
+
+"""
+    unsafe_modify!(p::Ptr{T}, op, x, [order::Symbol]) -> Pair
+
+These atomically perform the operations to get and set a memory address after applying
+the function `op`. If supported by the hardware (for example, atomic increment), this may be
+optimized to the appropriate hardware instruction, otherwise its execution will be
+similar to:
+
+    y = unsafe_load(p)
+    z = op(y, x)
+    unsafe_store!(p, z)
+    return y => z
+
+The `unsafe` prefix on this function indicates that no validation is performed on the
+pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
+that referenced memory is not freed or garbage collected while invoking this function.
+Incorrect usage may segfault your program.
+
+This is compatible with modifying pointers declared with `_Atomic` and `std::atomic`
+type in C11 and C++23 respectively. An error may be thrown if there is not support for
+atomically modifying the Julia type `T`.
+
+!!! compat "Julia 1.10"
+     This function requires at least Julia 1.10.
+
+See also: [`modifyfield!`](@ref), [`modifyproperty!`](@ref)
+"""
+function unsafe_modify!(p::Ptr, op, x, order::Symbol=:not_atomic)
+    return atomic_pointermodify(p, op, x, order)
+end
+
+"""
+    unsafe_replace!(p::Ptr{T}, expected, desired,
+                  [success_order::Symbol, [fail_order::Symbol=success_order]) -> (; old, success::Bool)
+
+These atomically perform the operations to get and conditionally set a memory address to
+a given value. If supported by the hardware, this may be optimized to the appropriate
+hardware instruction, otherwise its execution will be similar to:
+
+    y = unsafe_load(p, fail_order)
+    ok = y === expected
+    if ok
+        unsafe_store!(p, desired, success_order)
+    end
+    return (; old = y, success = ok)
+
+The `unsafe` prefix on this function indicates that no validation is performed on the
+pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
+that referenced memory is not freed or garbage collected while invoking this function.
+Incorrect usage may segfault your program.
+
+This is compatible with loading and storing pointers declared with `_Atomic` and
+`std::atomic` type in C11 and C++23 respectively. An error may be thrown if there is not
+support for atomically loading and storing the Julia type `T`.
+
+!!! compat "Julia 1.10"
+     This function requires at least Julia 1.10.
+
+See also: [`replacefield!`](@ref), [`replaceproperty!`](@ref)
+"""
+function unsafe_replace!(p::Ptr{T}, expected, desired, success_order::Symbol=:not_atomic, fail_order::Symbol=success_order) where {T}
+    @inline
+    xT = desired isa T ? desired : convert(T, desired)
+    return atomic_pointerreplace(p, expected, xT, success_order, fail_order)
+end
```

Comment:
Are implicit conversions on pointer stores a good idea..?

Reply:
Just noticed that regular `unsafe_store!` also does that :pensive:

Reply:
Would your preference be that it just throws an error otherwise

Reply:
Well, if the `convert` doesn't work, the intrinsic checks the type anyway so it wouldn't really change anything error-wise other than making the user choose how to convert in all cases (and possibly giving a different error). I just think it's an unexpected conversion, so I though I'd ask :shrug:

Reply:
After some digging, seems like this is from way back when in 2012, in f880afdbc25aacfb6cca970fd451741329fc089f by @vtjnash - it's a style thing I guess, it's just unexpected to have that conversion there when pretty much everything else about `unsafe_*` requires the user to make sure their code is already correct.

Reply:
My reasoning is this - JET can't find this, unless you turn on `:sound` mode, which can often give false positives IIRC:

```
julia> struct MyInt
           a::Int
       end

julia> Base.convert(::Type{Int}, a::MyInt) = a

julia> a = [2]
1-element Vector{Int64}:
 2

julia> f(a) = unsafe_store!(pointer(a), MyInt(1))
f (generic function with 1 method)

julia> JET.@report_call f(a)
No errors detected


julia> f(a)
ERROR: TypeError: in pointerset, expected Int64, got a value of type MyInt
Stacktrace:
 [1] unsafe_store!(p::Ptr{Int64}, x::MyInt) (repeats 2 times)
   @ Base ./pointer.jl:126 [inlined]
 [2] f(a::Vector{Int64})
   @ Main ./REPL[10]:1
 [3] top-level scope
   @ REPL[12]:1

julia> JET.@report_call mode=:sound f(a)
═════ 1 possible error found ═════
┌ @ REPL[10]:1 unsafe_store!(pointer(a), MyInt(1))
│┌ @ pointer.jl:126 #self#(p, x, 1)
││┌ @ pointer.jl:126 Base.pointerset(p, convert(T, x), Base.Int(i), 1)
│││ this builtin function call may throw: Base.pointerset(p::Ptr{Int64}, convert(T::Type{Int64}, x::MyInt)::MyInt, Base.Int(i::Int64)::Int64, 1)
││└──────────────────
```

If this type equality were required before calling the intrinsic, these kinds of mistakes would be found by JET without having to make sure the whole codebase is `:sound` :shrug:

Reply:
Should I change it to this
```julia
unsafe_store!(p::Ptr{T}, x, i::Integer=1) where {T} = unsafe_store!(p, convert(T, x), Int(i))
unsafe_store!(p::Ptr{T}, x::T, i::Int=1) where {T} = pointerset(p, x, i, 1)
```

Reply:
I'd personally not have the `convert` at all, but that's not my call to make :shrug:

Reply:
I'm assuming that would be breaking?

Reply:
Technically by SemVer I don't think so (the `convert` is undocumented), but it'd be good to have this discussion some place other than this PR, to not hold it up too much :) As long as the `convert` in your atomic bindings is also undocumented, it should be fine to remove at a later date together with removing it in regular `unsafe_store!`.

If we do want to resolve this with this PR (though that definitely requires more input than just mine here), it'd be good to make it consistent, i.e. such that `unsafe_modify!` also calls `convert` in a similar manner.

Reply:
Made a new issue here https://github.com/JuliaLang/julia/issues/49826